src/ctrans.c

```diff
@@ -320,142 +354,282 @@ int matmulti(_float *val, int n, _float mat[][3], int skip){
+    PyErr_SetString(PyExc_ValueError, "Dimension 1 of array must be 4");
+    goto error;
+  }
+
   dims[0] = grid_nsteps[0];
   dims[1] = grid_nsteps[1];
   dims[2] = grid_nsteps[2];
 
-  gridout = PyArray_ZEROS(3, dims, NPY_DOUBLE, 0);
+  gridout = (PyArrayObject*)PyArray_SimpleNew(3, dims, NPY_DOUBLE);
   if(!gridout){
-    goto cleanup;
+    PyErr_SetString(PyExc_MemoryError, "Could not allocate memory (gridout)");
+    goto error;
   }
-  Nout = PyArray_ZEROS(3, dims, NPY_ULONG, 0);
+
+  Nout = (PyArrayObject*)PyArray_SimpleNew(3, dims, NPY_ULONG);
   if(!Nout){
-    goto cleanup;
+    PyErr_SetString(PyExc_MemoryError, "Could not allocate memory (Nout)");
+    goto error;
   }
-  standarderror = PyArray_ZEROS(3, dims, NPY_DOUBLE, 0);
-  if(!standarderror){
-    goto cleanup;
+
+  stderror = (PyArrayObject*)PyArray_SimpleNew(3, dims, NPY_DOUBLE);
+  if(!stderror){
+    PyErr_SetString(PyExc_MemoryError, "Could not allocate memory (stderror)");
+    goto error;
   }
-  
-  n_outside = c_grid3d(PyArray_DATA(gridout), PyArray_DATA(Nout),
-		       PyArray_DATA(standarderror), PyArray_DATA(gridI),
-		       grid_start, grid_stop, data_size, grid_nsteps, norm_data);
-  
+
+  meanout = (PyArrayObject*)PyArray_SimpleNew(3, dims, NPY_DOUBLE);
+  if(!meanout){
+    PyErr_SetString(PyExc_MemoryError, "Could not allocate memory (meanout)");
+    goto error;
+  }
+ 
+  int retval;
+  retval = c_grid3d((double*)PyArray_DATA(gridout), (unsigned long*)PyArray_DATA(Nout),
+                    (double*)PyArray_DATA(meanout), (double*)PyArray_DATA(stderror), 
+                    (double*)PyArray_DATA(gridI), &n_outside,
+		                grid_start, grid_stop, (unsigned long)data_size, grid_nsteps, 1);
+
+  if(retval){
+    // We had a runtime error
+    PyErr_SetString(PyExc_RuntimeError, "Gridding process failed to run");
+    goto error;
+  }
+
   Py_XDECREF(gridI);
-  return Py_BuildValue("NNNl", gridout, Nout, standarderror, n_outside);
-  
- cleanup:
+  return Py_BuildValue("NNNNl", gridout, meanout, Nout, stderror, n_outside);
+
+error:
   Py_XDECREF(gridI);
   Py_XDECREF(gridout);
+  Py_XDECREF(meanout);
   Py_XDECREF(Nout);
-  Py_XDECREF(standarderror);
+  Py_XDECREF(stderror);
   return NULL;
 }
 
-unsigned long c_grid3d(double *dout, unsigned long *nout, double *standarderror, double *data,
-		       double *grid_start, double *grid_stop, int max_data,
-		       int *n_grid, int norm_data){
-  int i;
-  unsigned long *nout_ptr;
-  double *dout_ptr;
-  double *data_ptr;
-  
-  double *Mk = NULL;
-  double *Qk = NULL;
-  int grid_size = 0;
-
-  double pos_double[3];
-  double grid_len[3], grid_step[3];
-  int grid_pos[3];
-  int pos = 0;
-  unsigned long n_outside = 0;
+int c_grid3d(double *dout, unsigned long *nout, double *mout,
+             double *stderror, double *data, unsigned long *n_outside,
+             double *grid_start, double *grid_stop, unsigned long max_data,
+             unsigned long *n_grid, int norm){
```

Comment:
It will be good if several lines of doc can be added to explain what this function does.